File: sustain/scripts/build_cross_street_coords.py
import json
import time
from pathlib import Path

import pandas as pd
import requests

BASE_DIR = Path(__file__).resolve().parent.parent
EXCEL_PATH = BASE_DIR / "public" / "data" / "Efficiency Navigator Program - Data Support Group (2).xlsx"
OUTPUT_PATH = BASE_DIR / "public" / "data" / "cross_street_coords.json"

TARGET_SHEETS = [
    "OEI by Measure",
    "CDBG and ARPA by Measure",
    "Madison Capital 2024",
    "Madison Capital & EECBG 2025",
]

def normalize_text(value):
    text = str(value or "").strip()
    if text.lower() == "nan":
        return ""
    return " ".join(text.split())

# manual corrections for known problematic street names
def clean_street_name(street):
    street = normalize_text(street)

    corrections = {
        "Willage Green Lane E": "Village Green Lane E",
        "Manona Dr.": "Monona Dr.",
        "W. Fieldler Ln.": "W. Fiedler Ln.",
        "Wyldewood Dr. / Monterey Drive": "Wyldewood Dr. / Monterey Dr.",
    }

    for bad, good in corrections.items():
        street = street.replace(bad, good)

    return street

# tries to find the most likely header for cross streets
def resolve_cross_street_column(columns):
    for col in columns:
        if "cross street" in str(col).strip().lower():
            return col
    raise KeyError(f"Could not find Cross Street column. Available columns: {list(columns)}")

# parse excel workbook and extract relevant data
def collect_unique_cross_streets(excel_path):
    xls = pd.ExcelFile(excel_path)
    streets = set()

    for sheet_name in TARGET_SHEETS:
        if sheet_name not in xls.sheet_names:
            continue

        df = pd.read_excel(excel_path, sheet_name=sheet_name)
        cross_col = resolve_cross_street_column(df.columns)

        current_street = ""
        for raw_value in df[cross_col].tolist():
            street = normalize_text(raw_value)

            if street:
                current_street = street

            if current_street and current_street.lower() != "nan":
                streets.add(current_street)

    return sorted(streets)

# since street names in the excel are in a weird format
# this builds a list of query strings to try for geocoding a given street name
def build_queries(street):
    street = clean_street_name(street)

    queries = [f"{street}, Madison WI"]

    if "/" in street:
        a, b = [part.strip(" .") for part in street.split("/", 1)]
        queries.extend([
            f"{a} and {b}, Madison WI",
            f"{a} & {b}, Madison WI",
            f"intersection of {a} and {b}, Madison WI",
            f"{a}, {b}, Madison WI",
        ])

    return queries

# geocode a single street name using Nominatim API
def geocode_single_street(session, street):
    url = "https://nominatim.openstreetmap.org/search"
    params = {
        "format": "jsonv2",
        "limit": 1,
        "q": f"{street}, Madison WI",
    }

    response = session.get(url, params=params, timeout=30)
    response.raise_for_status()

    data = response.json()
    if not data:
        return None

    return {
        "lat": float(data[0]["lat"]),
        "lng": float(data[0]["lon"]),
    }

# geocode a street name, trying multiple query formats if needed
def geocode_street(session, street):
    street = clean_street_name(street)

    if "/" in street:
        parts = [part.strip(" .") for part in street.split("/", 1)]
        if len(parts) == 2:
            coord1 = geocode_single_street(session, parts[0])
            time.sleep(0.5)
            coord2 = geocode_single_street(session, parts[1])

            if coord1 and coord2:
                return {
                    "lat": (coord1["lat"] + coord2["lat"]) / 2,
                    "lng": (coord1["lng"] + coord2["lng"]) / 2,
                }

    return geocode_single_street(session, street)


def main():
    if not EXCEL_PATH.exists():
        raise FileNotFoundError(f"Excel file not found: {EXCEL_PATH}")

    streets = collect_unique_cross_streets(EXCEL_PATH)
    print(f"Found {len(streets)} unique cross streets")

    session = requests.Session()
    session.headers.update(
        {
            "User-Agent": "Sustain-Dane cross street geocoder/1.0 (local script)"
        }
    )

    coords = {}
    failures = []

    for i, street in enumerate(streets, start=1):
        try:
            coord = geocode_street(session, street)
            if coord is not None:
                coords[street] = coord
                print(f"[{i}/{len(streets)}] OK   {street} -> {coord}")
            else:
                failures.append(street)
                print(f"[{i}/{len(streets)}] MISS {street}")
        except Exception as e:
            failures.append(street)
            print(f"[{i}/{len(streets)}] FAIL {street} -> {e}")

        time.sleep(1.1)

    OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    with open(OUTPUT_PATH, "w", encoding="utf-8") as f:
        json.dump(coords, f, indent=2, ensure_ascii=False)

    print(f"\nSaved {len(coords)} coordinates to {OUTPUT_PATH}")

    if failures:
        print("\nFailed streets:")
        for street in failures:
            print(street)


if __name__ == "__main__":
    main()

File: sustain/scripts/build-cross-street-coords.mjs
/**
 * Regenerates public/data/cross_street_coords.json from the Efficiency Navigator workbook.
 * Run from repo: node scripts/build-cross-street-coords.mjs
 */
import * as XLSX from "xlsx";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import {
    geocodeCrossStreetIntersection,
    sleep,
} from "../src/lib/madisonGeocode.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(__dirname, "..");

const TARGET_SHEETS = [
    "OEI by Measure",
    "CDBG and ARPA by Measure",
    "Madison Capital 2024",
    "Madison Capital & EECBG 2025",
];

function normalizeText(value) {
    return String(value ?? "").replace(/\s+/g, " ").trim();
}

function resolveCrossStreetHeader(headers) {
    for (const header of headers) {
        const lower = String(header).toLowerCase().trim();
        if (lower.includes("cross street")) return header;
        if (lower.includes("cross") && lower.includes("street")) return header;
    }
    return null;
}

function collectStreetsFromWorkbook(wb) {
    const streets = new Set();
    for (const sheetName of TARGET_SHEETS) {
        const sheet = wb.Sheets[sheetName];
        if (!sheet) continue;
        const rows = XLSX.utils.sheet_to_json(sheet, {
            defval: "",
            raw: true,
        });
        if (!rows.length) continue;
        const headers = Object.keys(rows[0]);
        const crossH = resolveCrossStreetHeader(headers);
        if (!crossH) continue;
        let current = "";
        for (const row of rows) {
            const cell = normalizeText(row[crossH]);
            if (cell) current = cell;
            if (current) streets.add(current);
        }
    }
    return [...streets].sort();
}

const xlsxPath = path.join(
    root,
    "public",
    "Efficiency Navigator Program - Data Support Group.xlsx"
);
if (!fs.existsSync(xlsxPath)) {
    console.error("Missing workbook:", xlsxPath);
    process.exit(1);
}

const wb = XLSX.read(fs.readFileSync(xlsxPath), { type: "buffer" });
const list = collectStreetsFromWorkbook(wb);
console.log("Cross streets to geocode:", list.length);

const out = {};
for (let i = 0; i < list.length; i++) {
    const street = list[i];
    process.stdout.write(`[${i + 1}/${list.length}] ${street} … `);
    try {
        const c = await geocodeCrossStreetIntersection(street);
        if (c) {
            out[street] = c;
            console.log("OK", c.lat.toFixed(5), c.lng.toFixed(5));
        } else {
            console.log("MISS");
        }
    } catch (e) {
        console.log("ERR", e.message);
    }
    await sleep(280);
}

const outDir = path.join(root, "public", "data");
fs.mkdirSync(outDir, { recursive: true });
const outPath = path.join(outDir, "cross_street_coords.json");
fs.writeFileSync(outPath, JSON.stringify(out, null, 2));
console.log(
    "Wrote",
    Object.keys(out).length,
    "/",
    list.length,
    "to",
    path.relative(root, outPath)
);
